### PR TITLE
Fix ess imse update

### PR DESCRIPTION
**Pull request by bob-carpenter**
This PR:

* pulls autocorrelation calculations into their own file 
* adds more autocorrelation tests including exceptions
* adds more tests for effective sample size including independent draws
* adds more doc

It fixes issues #13 and #15.

### CODE REVIEW

File: bayes_kit/ess.py [hunk 1]
```diff
@@ -71,7 +23,7 @@ def first_neg_pair_start(chain: VectorType) -> IntType:
     while n + 1 < N:
         if chain[n] + chain[n + 1] < 0:
             return n
-        n = n + 2
+        n += 2
     return N
```
Comment (jsoules) [inline bayes_kit/ess.py:27]: Just to make sure this is working as intended:

If given [2, -1, -3, -5, 6, 12], should this function return 2 or 1?

As written it returns 2, since it always counts pairwise, but obviously -1 + -3 is -4 (< 0), and the docstring doesn't indicate that the pairs have to start on an even index.

**Comment by bob-carpenter** [comment]
This is ready to re-review.  I got up to 100% test coverage, which was really useful for making me be more careful defining and testing boundary conditions.  I renamed the "internal" function `_end_pos_pairs` after @magland pointed out the first name misspecified the function, then documented the boundary conditions, then fixed the function to match the tests.

**Comment by bob-carpenter** [comment]
Also, this will close issues #15 and #13.

File: bayes_kit/ess.py [hunk 2]
```diff
@@ -1,66 +1,26 @@
...
-    """
-    return autocorr_fft(chain)
-
-def first_neg_pair_start(chain: VectorType) -> IntType:
-    """
-    Return the index of first element of the sequence whose sum with the following
-    element is negative, or the length of the sequence if there is no such element.
+def _end_pos_pairs(chain: VectorType) -> IntType:
+    """Return the index of first element of the sequence whose sum with
+
```
Comment (jsoules) [inline bayes_kit/ess.py:11]: extra newline

File: bayes_kit/ess.py [hunk 3]
```diff
@@ -1,66 +1,26 @@
...
+    the following element is negative, or the length of the sequence if
+    there is no such element.  Equivalently, this is the upper bound for
+    the range of indexes that come in positive pairs starting from 0.
+    For exmaple, `_end_pos_pairs([1, -.5, .25, -.3]) == 2` because the
```
Comment (jsoules) [inline bayes_kit/ess.py:15]: typo (exmaple)

File: bayes_kit/ess.py [hunk 4]
```diff
@@ -1,66 +1,26 @@
...
-    autocorrelation estimates at all lags for the specified sequence
-    """
-    return autocorr_fft(chain)
-
-def first_neg_pair_start(chain: VectorType) -> IntType:
-    """
-    Return the index of first element of the sequence whose sum with the following
-    element is negative, or the length of the sequence if there is no such element.
+def _end_pos_pairs(chain: VectorType) -> IntType:
+    """Return the index of first element of the sequence whose sum with
```
Comment (jsoules) [inline bayes_kit/ess.py:10]: "Return the index of the first *even-indexed* element" maybe? It's implicit in the example and the comment about "indexes that come in positive pairs" but as written this sentence doesn't mention index parity, so a quick skim would give the wrong idea.

File: bayes_kit/ess.py [hunk 5]
```diff
@@ -1,66 +1,26 @@
...
+    = 4`, in the case where the second pair (.25, -.1) is positive.
+    Trailing odd elements are ignored, so `_end_pos_pairs([1, -.5, .25, -.3, .05]) == 2`
+    and `_end_pos_pairs([1, -.5, .25, -.1, .05]) = 4`.
     
     Parameters:
-    chain: input sequence
+    chain: input sequence (typically of autocorrelations)
 
     Return:
     index of first element whose sum with following element is negative, or
```
Comment (jsoules) [inline bayes_kit/ess.py:26]: As above, I think it'd be best to call out that these are even-indexed elements only (and the function can never return an odd number).

File: bayes_kit/ess.py [hunk 6]
```diff
@@ -85,15 +45,15 @@ def ess_ipse(chain: VectorType) -> FloatType:
...
     """
     if len(chain) < 4:
-        raise ValueError(f"ess requires len(chains) >=4, but {len(chain) = }")
+        raise ValueError(f"ess requires len(chains) >= 4, but {len(chain)=}")
     acor = autocorr(chain)
-    n = first_neg_pair_start(acor)
-    sigma_sq_hat = acor[0] + 2 * sum(acor[1:n])
-    ess = len(chain) / sigma_sq_hat
+    n = _end_pos_pairs(acor)
+    iat = 2 * acor[0:n].sum() - 1
```
Comment (jsoules) [inline bayes_kit/ess.py:55]: Is `iat` a standard notation that we'd expect everybody to recognize? I don't recognize it (which isn't in itself surprising) but it also didn't appear as such when I skimmed the Wikipedia article on sample size determination, I can't find corresponding notation in the Stan reference page on effective sample size, and a google search for "iat statistics" or "iat effective sample size" just pulls up a bunch of stuff about the Implicit Association Test (i.e. psychology), so I think it would be pedagogically helpful to use a more descriptive name (or at least comment it).

(I'm okay with `acor` since it's pretty spelled out by the name and documentation of the function call that populates it.)

Comment (jsoules) [reply bayes_kit/ess.py:55]: (Same note probably applies for the imse function below)

**Review by jsoules — COMMENTED** [review]
I had a couple proofreading notes, then some comments about comments.

File: bayes_kit/ess.py (hunk 6 above)
Comment (bob-carpenter) [reply bayes_kit/ess.py:55]: It's for "integrated autocorrelation time".  Here's a [blog post explaining](https://dfm.io/posts/autocorr/) where @dfm explains.

If "iat" is too short for a local variable name, could you please suggest the shortest thing you'll approve and I'll use that.

We could expose the IAT function itself.  Sometimes people like to report that, too.  Then the definition of ESS simplifies to `ESS = chain_length / IAT`.

File: bayes_kit/ess.py (hunk 2 above)
Comment (bob-carpenter) [reply bayes_kit/ess.py:11]: fixed.

File: bayes_kit/ess.py (hunk 5 above)
Comment (bob-carpenter) [reply bayes_kit/ess.py:26]: fixed along with last one

File: bayes_kit/ess.py (hunk 6 above)
Comment (jsoules) [reply bayes_kit/ess.py:62]: As with `acorr`, I'm not fussed if there's something more explicit right nearby, it was just as-is this wasn't enough for me to know the right things to google.

I really like the idea of exposing an `integrated_autocorr_time` function--it fixes the naming issue, breaks the math into semantically meaningful components, and also exposes some reusable code that people might want direct access to anyway. Three wins!

**Comment by WardBrian** [comment]
I'm not sure if you'd like to fix them here or in #10 or somewhere else, but I just wanted to note that `mypy --strict .` raises 61 errors in this branch. It's about 25 between `ess.py`, `test_ess.py` and `test_autocorrelation.py`, with the rest in the RHat code you don't really touch here

File: bayes_kit/ess.py (hunk 6 above)
Comment (bob-carpenter) [reply bayes_kit/ess.py:62]: I'll update and add an IAT function and tests.

**Comment by bob-carpenter** [comment]
I'l fix the mypy tests.  Sorry for skipping bits of the workflow.  

Would there be a way to automate (a) black formatting, (b) strict mypy testing, and (c) anything else that's going to cause a PR to be rejected if I forget to do it manually?

**Comment by WardBrian** [comment]
I plan on automating that as soon as our current `main` branch would pass it, but a lot of the type errors in these files are already checked in there so it wouldn’t be feasible yet

**Comment by bob-carpenter** [comment]
Another question on doc.  The extra line space is because emacs thinks the first line of doc should fit on a single line.  Should I be forcing the first sentence of doc to fit on a single line?  Is it OK if it runs over?  black didn't reformat it, but emacs doesn't like it.

**Comment by WardBrian** [comment]
I believe the standard is if it takes more than one line, you should add a newline after the opening `"""`

e.g.

```python

"""A single line docstring"""

"""
A docstring which is much longer, taking up 
multiple lines or even including multiple sections
"""
```

Black doesn't touch docstrings, but the python emacs package might have this opinion built-in

**Comment by bob-carpenter** [comment]
Thanks.  The examples for the Google-style doc start right after the `"""` but they're all one-liners.  I can try to be more terse if that's the standard in Python.

**Comment by bob-carpenter** [comment]
I just want to follow conventions which looks like terse first sentences.   That's what javadoc wanted, too.  I'll see what I can do.

File: bayes_kit/iat.py [hunk 7]
```diff
@@ -29,60 +32,56 @@ def _end_pos_pairs(chain: VectorType) -> IntType:
...
+    where `Z = {..., -2, -1, 0, 1, 2, ...}` is the set of integers.
 
-    Because the autocorrelations are symmetric, `ac[n] == ac[-n]` and
-    `ac[0] = 1`, this reduces to
+    Because the autocorrelations are symmetric, `autocorr[n] == autocorr[-n]` and
+    `autocorr[0] = 1`, the sum reduces to
 
     ```
-    IAT = -1 + 2 * SUM_{n in N} ac[n],
+    IAT = -1 + 2 * SUM_{n in N} autocorr[n],
```
Comment (jsoules) [inline bayes_kit/iat.py:69]: Just wanted to flag that this is really nice & helpful exposition. Thanks!

Comment (jsoules) [reply bayes_kit/iat.py:69]: Is it worth explicitly noting that the `-1` is because we double-counted `acor[0]`? It doesn't take much reading between the lines to figure it out, but you never know who's asleep at the wheel.

File: bayes_kit/iat.py [hunk 8]
```diff
@@ -115,43 +113,45 @@ def iat_imse(chain: VectorType) -> FloatType:
...
+    The IAT can be less than one in cases where the Markov chain is
+    anti-correlated.
 
     Parameters:
-        chain: Markov chains whose ESS is returned
+    chain: A Markov chain.
 
     Return:
-        estimated effective sample size for the specified Markov chain
+    The integrated autocorrelation time (IAT) fo the specified chain.
```
Comment (jsoules) [inline bayes_kit/iat.py:152]: typo (fo -> for)

File: bayes_kit/iat.py [hunk 9]
```diff
@@ -115,43 +113,45 @@ def iat_imse(chain: VectorType) -> FloatType:
...
 
 
 def iat(chain: VectorType) -> FloatType:
     """
-    Return an estimate of the effective sample size of the specified Markov chain
-    using the default ESS estimator (currently IMSE).  Evaluated by delegating
-    to `iat_imse()`.
+    Return an estimate of the integrated autocorrelation time (IAT)
+    of the specified Markov chain. Evaluated by delegating to the
+    initial monotone sequence estimator, `iat_imse(chain)`.
```
Comment (jsoules) [inline bayes_kit/iat.py:143]: Would there be any reason (performance?) someone might want the `ipse` instead? Is it worth exposing the ability to choose that? (I'm guessing not really)

File: test/test_iat.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    E_iat = 1.0
+    N = 20_000
+    y = np.random.normal(size=N)
+    hat_iat = bk.iat(y)
+    np.testing.assert_allclose(E_iat, hat_iat, rtol=0.1)
+
+def test_iat_ar1():
+    # tests correlated (rho > 0), uncorrelated (rho = 0), anti-correlated (rho < 0)
+    for rho in np.arange(-0.5, 0.5, step=0.2):
+        run_iat_ar1_test(rho, 20_000)
```
Comment (jsoules) [inline test/test_iat.py:46]: Not to be pedantic, but with stepsize 0.2, this series will skip right over the case where rho = 0--it hits -0.5, -0.3, -0.1, 0.1, 0.3.
Do we care?

Comment (bob-carpenter) [reply test/test_iat.py:46]: The doc should match the code, so this isn't pedantic at all.  Fixed.

File: bayes_kit/iat.py (hunk 9 above)
Comment (bob-carpenter) [reply bayes_kit/iat.py:143]: I'm OK getting rid of IPSE; even though it's a bit faster than IMSE, it has an overestimation bias.    Do you think I should get rid of it?

I still think we should keep the `ess_imse` and `iat_imse` definitions and delegate `ess` and `iat`.  I know at least for ess we will have further estimators that work cross chain.  The other option is to not provide a default implementation.  I'm always on the fence about choosing default implementations for users, but in the end, I think they're going to want o see `ess()` or `iat()` in most applied code.

Comment (jsoules) [reply bayes_kit/iat.py:144]: I don't have a strong feeling about IPSE--it's there, it's implemented, it works; no reason to discard that work, especially if we might want to expose it later. The counterargument is that it's more surface area, and unreachable unused code gets out of date (not that any code in Python is ever truly unreachable).

Unfortunately, I'm kind of clueless whether anybody would in practice actually want to have a choice here.

And if we expose that choice, however we did so, I could imagine a tricky situation where we wind up having some deeply nested calls such that there's an inconsistency between the implementations used in different places--whether because it's hard to make sure the "use-this-implementation" flag gets passed around everywhere `ess` is called, or because users call `ess_ipse` in their own direct code but our version doesn't. Of course, again, dunno if such a scenario is actually likely in practice.

I guess it would come down to three questions:
- Would anybody have a reason to use IPSE?
- Are these two the only implementations worth implementing, or could we imagine adding more later?
- Are these used only/mainly for evaluating sample quality after the fact, or are there cases where a sampling method might want to make use of these in a way that affects its behavior?

Comment (bob-carpenter) [reply bayes_kit/iat.py:144]: I talked to @WardBrian and concluded we should leave it in for pedagogical purposes.  

* I think the utility for IPSE is pedagogical and I don't think anyone would use one in an application unless they really needed the extra speed.  The standard way to teach these estimators is to first describe the initial positive sequence method, which was in common use until IPSE was developed, then add a monotonicity condition on top of that.  
* Yes, we'll probably be doing other versions of these eventually, such as a Bayesian estimator that makes a parametric or non-parametric assumption about the shape of the autocorrelation vs. lag curve---it's an active area of research.
* Yes, we will eventually want to use ESS at run time to set a minimum ESS target and run until it's hit (it won't affect the transitions, just when we stop)

Comment (jsoules) [reply bayes_kit/iat.py:144]: Sounds good to me. All of this points to the same conclusion--leave it in for pedagogy (and so people could technically call it if they really wanted to), but have a standard pathway that delegates to a specific implementation.

**Review by jsoules — APPROVED** [review]
This has been open for a while, but reviewing it again this morning I have nothing further on it and would be happy for it to be merged.
